### PR TITLE
Added option for formatting and arbitrary metadata

### PR DESCRIPTION
**Pull request by Mityo**
@nemtsov please review and let me know if you have any remarks.
Decided to leave the message data out of the format string in order to let console methods handle it.
As far as performance go my tests show that this implementation is slightly faster  than the current one.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -30,6 +30,16 @@ logger.info('Hello %s!', 'world');
 
 - `createLoggers()` takes a minimum logging `level`, which is one of: `['info', 'warn', 'error']` and defaults to `info`. For example if you select `warn`, all `info` logs will be ignored. `warn` and `error` logs will be displayed.
 
+Support for formatting and arbitrary metadata
+```js
+import createLoggers from 'namespaced-console-logger';
+
+const loggers = createLoggers('info', '${timestamp} ${hostname} (${namespace}) ${level}:', { hostname: 'machinename' });
```
Comment (nemtsov) [inline README.md]: I'd either use `{{` and `}}` (like in mustache) or `:timestamp` like in [morgan](https://github.com/expressjs/morgan#combined) as variable delimiters to (a) avoid confusing these with string interpolation in template literals and (b) to avoid needing to escape these in template literals, (e.g., ``` `${expression} \${not_an_expression}` ```.

Comment (nemtsov) [reply README.md]: Then I'd dump `meta` in favor of putting the value directly into the template string:

So, instead of:
```
'${timestamp} ${hostname} (${namespace}) ${level}:', { hostname: 'machinename' }
```

I'd do:
```
${timestamp} machinename (${namespace}) ${level}:'
```

File: src/index.js [hunk 2]
```diff
@@ -7,32 +9,38 @@ const allLevels = ['error', 'warn', 'info'];
  * logging level of `info`.
  *
  * @param {String} level (info|warn|error) minimum logging level
+ * @param {String} format format meta data
+ * @param {Object} meta meta data
  */
 
-export default function (level = 'info') {
+export default function (level = 'info', format, meta) {
```
Comment (nemtsov) [inline src/index.js:16]: Like I mentioned above, I don't think there's a reason for `meta` when we have `format`, unless we handle functions as values of meta keys.

File: src/template.js [hunk 3]
```diff
@@ -0,0 +1,28 @@
+/**
+ * Parse format string and return object
+ *
+ * @param {String} format format string
+ */
+export default function (format) {
```
Comment (nemtsov) [inline src/template.js]: Great idea to make a function from the provided format!

File: src/template.js [hunk 4]
```diff
@@ -0,0 +1,28 @@
+/**
+ * Parse format string and return object
+ *
+ * @param {String} format format string
+ */
+export default function (format) {
+  let regex = /\$\{(\w+)\}/g;
+  let result = [];
```
Comment (nemtsov) [inline src/template.js]: nit: `regex` and `result` should be `const` since they don't change

File: src/template.js [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+    result.push('(data.' + match[1] + ' || \'' + match[0] + '\')');
+
+    index = match.index;
+    length = match[0].length;
+  }
+  if ((index + length) < format.length) {
+    result.push('\'' + format.substring(index + length, format.length) + '\'');
+  }
+
+  let funcBody = 'return ' + result.join(' + ') + ';';
```
Comment (nemtsov) [inline src/template.js]: nit: `const funcBody`

File: test/template-test.js [hunk 6]
```diff
@@ -0,0 +1,37 @@
+import template from '../src/template'
+
+
+describe('template', () => {
+  it('execute', (done) => {
```
Comment (nemtsov) [inline test/template-test.js]: `done` is only necessary if the test is asynchronous. You can remove it here and in the tests below.

Comment (nemtsov) [reply test/template-test.js]: nit: In order to make the output readable, use:

 `it('should execute',` that way you can read it as `template should execute` (where the `it` is replaced by the previous `describe` and the rest can be read.

File: test/template-test.js [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+    let format = '${timestamp} (${namespace}) ${level}: ${data}';
```
Comment (nemtsov) [inline test/template-test.js]: for this variable and other below, if the variable doesn't change, please use `const`

File: test/template-test.js [hunk 8]
```diff
@@ -0,0 +1,37 @@
...
+
+    let str = compiled(data);
+    str.should.be.equal(data.timestamp + ' (' + data.namespace + ') ' + data.level + ': ' + data.data);
+    
+    delete data['namespace'];
+    str = compiled(data);
+    str.should.be.equal(data.timestamp + ' (${namespace}) ' + data.level + ': ' + data.data);
+    
+    done();
+  })
```
Comment (nemtsov) [inline test/template-test.js]: nit: missing semicolon and a newline

File: test/template-test.js [hunk 9]
```diff
@@ -0,0 +1,37 @@
...
+  it('execute with no literals', (done) => {
+    let format = 'Message: Hellow world';
+    let data = {timestamp: new Date().toISOString(), namespace: 'class', level: 'INFO'};
+    let compiled = template(format);
+
+    let str = compiled(data);
+    str.should.be.equal('Message: Hellow world');    
+    done();
+  })
```
Comment (nemtsov) [inline test/template-test.js]: These tests are missing some edge cases:

----
```
> template('')()
undefined          // should be '' instead of undefined
```

----

```
> template('${v}')()
TypeError: Cannot read property 'v' of undefined
```

but:
```
> template('${v}')({})
'${v}'
```

Is that expected? Or should there be a default empty object? Either way, we should have tests for these edge cases.

**Review by nemtsov — CHANGES_REQUESTED** [review]
Great work! I especially like creating a function from the template for performance. Some feedback below; mostly minor.

File: src/index.js (hunk 2 above)
Comment (Mityo) [reply src/index.js:16]: kept the meta and added support for function as a key value

**Comment by Mityo** [comment]
@nemtsov, implemented the requested changes and added support for function in meta.

File: src/template.js [hunk 10]
```diff
@@ -0,0 +1,33 @@
...
+
+  let index = 0;
+  let length = 0;
+  let match;
+  while (match = regex.exec(format)) {
+    if (index + length < match.index) {
+      result.push('\'' + format.substring(index + length, match.index) + '\'');
+    }
+    //fast check for a function: !!(object && object.constructor && object.call && object.apply) 
+    result.push('((!!(data.' + match[1] + ' && data.' + match[1] + '.constructor && data.' + match[1] + '.call && data.' + match[1] + '.apply) ? data.' + match[1] + '() : data.' + match[1] + ') || \'' + match[0] + '\')');
```
Comment (nemtsov) [inline src/template.js]: instead of determining whether it's a function at run-time, can you pass the `meta` into the template function and determine whether it's a function by using the value type?

Comment (Mityo) [reply src/template.js]: I can but it's insignificant change, this check is really fast.

File: src/index.js (hunk 2 above)
Comment (Mityo) [reply src/index.js]: Done!

File: src/template.js (hunk 10 above)
Comment (Mityo) [reply src/template.js]: Done!

**Review by nemtsov — APPROVED** [review]

**Comment by nemtsov** [comment]
Great work, @Mityo! Thank you for your contribution.

**Comment by nemtsov** [comment]
@Mityo, this PR landed in version `1.1.0`.
